[DoctrinBridge] make Uid types stricter

nicolas-grekas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.2
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Reviewing #38600 made me realize we don't need to deal with converting strings to db values.
We should only support converting actual `AbstractUid` instances in the DB.
Also, the binary types should *not* extend `GuidType`.

nicolas-grekas committed:
[DoctrinBridge] make Uid types stricter

stof commented:
> Also, the binary types should _not_ extend `GuidType`.

IMO, none of them should, as they change the PHP format

nicolas-grekas commented:
> IMO, none of them should, as they change the PHP format

Alright, updated. I also removed `requiresSQLCommentHint()` because I don't think comment-hints are required.

stof commented:
>  I also removed `requiresSQLCommentHint()` because I don't think comment-hints are required.

they are required, because the type does not register a mapping type, so schema introspection would not recognize them without the special comment

nicolas-grekas commented:
Thanks, reverted this part :)

fabpot commented:
Thank you @nicolas-grekas.

mvorisek commented on src/Symfony/Bridge/Doctrine/Tests/Types/UlidTypeTest.php:
```diff
@@ -44,7 +44,7 @@ protected function setUp(): void
     }
 
-    public function testUlidConvertsToDatabaseValue(): void
+    public function testUlidConvertsToDatabaseValue()
```
why to remove the `void` return type?

nicolas-grekas replied:
Because test cases don't define an API; they don't need to declare their signature. In the end, it's pure visual noise.

nicolas-grekas commented:
(Partially reverted in #38986)
